### PR TITLE
Inline cloud-init user data

### PR DESCRIPTION
**Pull request by cynepco3hahue**
Make possible to add cloud-init data to VM without
encode it to base64 format.

Signed-off-by: Lukianov Artyom <alukiano@redhat.com>

### CODE REVIEW

**Comment by cynepco3hahue** [comment]
@rmohr @davidvossel @fabiand Hi guys, I want to know your opinion
- do we want to support `UserDataBase64` field?
- do we want to suppot k8s secret user-data?
From my experience with cloud-init under oVirt and OpenStack project it used to make some initial VM's configuration and must not include some vulnerable data, so I do not sure that we need to use k8s secret.
But it still can be useful to have some cloud-init configuration that number of VM's can use it only by specify configuration name under VM `cloudInitNoCloud` source.
IMHO `ConfigMap` more suitable for it.
- do we want to support merge of user-data files(in the case when we have two sources of user-data)?
I know cloud-init supports something like this https://cloudinit.readthedocs.io/en/latest/topics/merging.html but I do not sure on what stage it applied and if it will work with generated cloud-init ISO image.

File: cluster/vm-nocloud.yaml [hunk 1]
```diff
@@ -21,7 +21,12 @@ spec:
   volumes:
     - name: registryvolume
       registryDisk:
-        image: kubevirt/cirros-registry-disk-demo:devel
+        image: kubevirt/fedora-cloud-registry-disk-demo:devel
```
Comment (rmohr) [inline cluster/vm-nocloud.yaml]: Once you add a test, make sure that you use cirros and bash scripts instead of fedora. It is unbearably slow.

**Comment by rmohr** [comment]
On Thu, Feb 22, 2018 at 9:27 AM, Artyom Lukianov <notifications@github.com>
wrote:

> @rmohr <https://github.com/rmohr> @davidvossel
> <https://github.com/davidvossel> @fabiand <https://github.com/fabiand> Hi
> guys, I want to know your opinion
>
>    - do we want to support UserDataBase64 field?
>
> Yes. For automation it is in my opinion more intuitive to use. Also if
such an init-script is long it can be hard to read the vm definition.


>
>    - do we want to suppot k8s secret user-data?
>
> Yes definitely. We store the data as secure as k8s stores it. Also we
provide it to the vm equally secure like k8s provides it to containers.
Engineers can use it right-away to mix Pod- and VM- components of an
application. For Pods they can provide the app credentials via k8s secrets,
for the vm components they can provide it through cloud-init via a k8s
secret.

>
>    - From my experience with cloud-init under oVirt and OpenStack project
>    it used to make some initial VM's configuration and must not include some
>    vulnerable data, so I do not sure that we need to use k8s secret.
>
>
I understand where that is coming from. However, I would argue that we can
follow the k8s pattern here, that secrets are the preferred built-in way to
pass sensible information to applications (one can always opt out and
simply not use it). I would even expect that we start soon on projecting
secrets and config maps into cloud-init disks (see
https://kubernetes.io/docs/concepts/storage/volumes/ for context).


>    - But it still can be useful to have some cloud-init configuration
>    that number of VM's can use it only by specify configuration name under VM
>    cloudInitNoCloud source.
>    IMHO ConfigMap more suitable for it.
>    - do we want to support merge of user-data files(in the case when we
>    have two sources of user-data)?
>    I know cloud-init supports something like this
>    https://cloudinit.readthedocs.io/en/latest/topics/merging.html
>    <https://cloudinit.readthedocs.io/en/latest/topics/merging.html> but I
>    do not sure on what stage it applied and if it will work with generated
>    cloud-init ISO image.
>
>
Maybe we can use it to achieve some kind of k8s-like projections? Not sure
about that. Need to understand this in more detail first.

Roman

**Comment by davidvossel** [comment]
Lets keep the k8s secret and base64 logic. This new logic of having the plan text string in the manifest is great for simple cloud-init configs and I'm glad it's being added. As for merging of cloud-init, I think we need a strong use case before considering introducing that logic.

**Comment by cynepco3hahue** [comment]
@rmohr @davidvossel In this case, can we just be sure in order of usage, currently we have priority from high to low
- `UserDataSecretRef`
- `UserData`
- `UserDataBase64`
If `UserDataSecretRef` is defined it will override `UserData` content.
If both `UserData` and `UserDataBase64` is defined it will take content only from `UserData`

Does it ok? I just need do not forget to specify it under user-guide :smile:

**Comment by davidvossel** [comment]
> In this case, can we just be sure in order of usage, currently we have priority from high to low

Well, actually we should probably follow the same pattern we use for VolumeSource, where only one value is allowed to be set on the CloudInitNoCloudSource struct.

**Comment by rmohr** [comment]
> Well, actually we should probably follow the same pattern we use for VolumeSource, where only one value is allowed to be set on the CloudInitNoCloudSource struct.

Agreed that is the intention. We really need to start adding the admission webhook for validations and defaulting.

**Comment by cynepco3hahue** [comment]
Ok I will check where can I add validation

**Comment by rmohr** [comment]
See also #556

**Comment by davidvossel** [comment]
> Ok I will check where can I add validation

just in case it isn't obvious. We don't have the validation webhook implemented yet.

**Comment by cynepco3hahue** [comment]
@davidvossel 
> Well, actually we should probably follow the same pattern we use for VolumeSource, where only one value is allowed to be set on the CloudInitNoCloudSource struct.

I checked `VolumeSource` and I can not find any validation. Can you point me?

**Comment by rmohr** [comment]
> I checked VolumeSource and I can not find any validation. Can you point me?

As mentined by @davidvossel we don't have any validation in place right now, since we don't yet use the webhook. We are just talking about the principle.

**Comment by cynepco3hahue** [comment]
@rmohr I thought we do not have webhook validation, but still have some validation under our code :smile: 
So I will leave this PR as it now, and when we will have validation we will need add cloud-init to it.

**Comment by rmohr** [comment]
@cynepco3hahue when we had our own apiserver we had validation.

**Comment by cynepco3hahue** [comment]
@rmohr @davidvossel Do you have some additional comments or we can run CI on it?

**Comment by rmohr** [comment]
ok to test

**Comment by rmohr** [comment]
@cynepco3hahue make sure to addd an example to the user-guide

**Pull request merged**